ci/funs.sh

```diff
@@ -1,8 +1,81 @@
-# utilities used in CI pipelines to avoid duplication.
+# Utilities used in CI pipelines and tooling to avoid duplication.
+# Prefer nim scripts whenever possible.
+# functions starting with `_` are considered internal, less stable.
 
 echo_run () {
   # echo's a command before running it, which helps understanding logs
   echo ""
   echo "$@"
   "$@"
 }
+
+nimDefineVars(){
+  nim_csourcesDir=csources_v1 # where we clone
+  nim_csourcesUrl=https://github.com/nim-lang/csources_v1.git
+  nim_csourcesHash=a8a5241f9475099c823cfe1a5e0ca4022ac201ff
+  nim_csources=bin/nim_csources_$nim_csourcesHash
+}
+
+_nimNumCpu(){
+  # linux: $(nproc)
+  # FreeBSD | macOS: $(sysctl -n hw.ncpu)
+  # OpenBSD: $(sysctl -n hw.ncpuonline)
+  # windows: $NUMBER_OF_PROCESSORS ?
+  echo $(nproc 2>/dev/null || sysctl -n hw.logicalcpu 2>/dev/null || getconf _NPROCESSORS_ONLN 2>/dev/null || 1)
+}
+
+_nimBuildCsourcesIfNeeded(){
+  if [ "${NIM_CSOURCES_USE_BUILD:-0}" == "1" ]; then
```

Comment:
Remove this. Instead it could detect if `make` is available and reliable.

Reply:
the problem is the arguments are different:
* makefile uses thse: `CC=gcc ucpu=arm64`
* build.sh uses these: `--cpu amd64 --os darwin`

the CI pipelines have been simplified to simply call things like `nimBuildCsourcesIfNeeded CC=gcc ucpu=arm64`; falling back to build.sh would pass arguments that can't be processed.

It worked before this PR because each CI copied what's now done inside `nimBuildCsourcesIfNeeded`, which obviously wasn't ideal.

I'm ok with any approach that doesn't require duplicating the logic outside of ci/funs.sh

Reply:
But no script actually sets `NIM_CSOURCES_USE_BUILD` anyway, it looks like dead code. I don't understand what it has to do with argument forwarding. `build.sh` could also acccept make-like arguments if that's not too much effort.

Reply:
> I don't understand what it has to do with argument forwarding. 

if you call `nimBuildCsourcesIfNeeded CC=gcc ucpu=arm64` and neither make or gmake is available, calling `build.sh "$@"` won't work because it expects arguments like `--cpu amd64 --os darwin`, not like `CC=gcc ucpu=arm64`

here's an option: given how ubiquitous it is to have at least make or gmake available, how about i remove these lines:
```
if [ "${NIM_CSOURCES_USE_BUILD:-0}" == "1" ]; then ( echo_run cd $nim_csourcesDir; echo_run sh build.sh "$@" )
```

and just bail out if make or gmake is unavailable; and then future work may add support for the same arguments as makefile in build.sh to avoid such complications, in case we need to still support build.sh when makefile is not available

Reply:
Fine with me.

Reply:
done